sklearn/pipeline.py

```diff
@@ -531,6 +531,36 @@ def _pairwise(self):
         # check if first estimator expects pairwise input
         return getattr(self.steps[0][1], '_pairwise', False)
 
+    def get_feature_names(self, input_features=None):
+        """Get feature names for transformation.
+
+        Transform input features using the pipeline.
+        If the last step is a transformer, it's included
+        in the transformation, otherwise it's not.
+
+        Parameters
+        ----------
+        input_feature : array-like of string
+            Input feature names.
+
+        Returns
+        -------
+        feature_names : array-like of string
+            Transformed feature names
+        """
+        feature_names = input_features
+        with_final = hasattr(self._final_estimator, "transform")
```

Comment:
do you have an alternative that's not slicing?

Reply:
The other option would be to give all supervised models a pass-through ``get_feature_names`` maybe?

Reply:
yeah I also think it's a bit confusing. But the most common use-case should be syntactically nice.
Maybe like ``pipe.pop()`` or something.
Have you thought about slicing cloned pipelines?
The "easiest" way I could think of was freezing all steps, making the sliced pipeline immutable. That's probably what we want, right? Creating a clone of the sliced pipeline in case someone wants to refit it should be pretty simple, right?

I didn't think doing this would link to slicing pipelines and freezing. I thought this one was the easy one :-/ Though I guess we really need the freezing "only" once we have the cloning pipeline. So it might be possible to implement this without solving the other issues first?

Reply:
It does link to slicing pipelines, but doesn't need to relate to freezing; slicing the pipeline should make something like a view of that slice, regardless of behaviour under cloning. Also, raising an error with non-transformers at the end of a pipeline is perfectly good for a start.

Reply:
> Also, raising an error with non-transformers at the end of a pipeline is perfectly good for a start.

I think that would be a very ugly solution, because it requires the user to either specify the pipeline in a weird way or to slice after the fact.

Would you allow fitting on a "view" of that slice? That might lead to unexpected consequences - though I guess not if cloning happens in ``fit``.

Reply:
See https://github.com/scikit-learn/scikit-learn/pull/8448#issuecomment-442206329

Reply:
@jnothman Can you explain a bit more why you find this controversial? 
Is it the way we the last estimator is skipped, or is it the actual skipping? In case of the second one: you think it should be an explicit action of the user to slice his full pipeline to remove the last estimator, to know the feature names of the features used in the last estimator?

Reply:
My understanding is that @jnothman would like the user to explicitly skip the last estimator, say via slicing. I think that actually makes more logical sense.

There is really no way for the pipeline to know whether the user meant to include the last model or not. For example LinearDiscriminantAnalysis has a transform and a predict, same for KMeans, and the pipeline doesn't know if the user is using predict on the pipeline (in which case they wanted the input to the last step) or transform (in which case they wanted the output of the last step).

Reply:
I addressed this comment, now that we have slicing ;)